### PR TITLE
MSP Updates: current, heading, and stick normalization

### PR DESCRIPTION
**Pull request by dustin**
The MSP stuff was previously mostly bench tested.  There were a few things that I noticed when I actually tried flying it for the first time as well as having tried an s.bus rx for the first time in quite a while.

* current values are off by an order of magnitude with MWOSD configured as baseflight.  For multiwii, it'd probably be right, but more resolution seems better
* Heading was not implemented.  Turns out it was rather trivial.
* s.bus stick values sent directly to MWOSD makes all the values unusable in the OSD.  Need to change how sticks are sent.

### CODE REVIEW

**Comment by tracernz** [comment]
Also noticed this when I had it connected to Baseflight Configurator. LGTM.

**Comment by dustin** [comment]
I tested this on the bench, and then flew with it last night.

**Comment by dustin** [comment]
I added the work I did on heading and the stick updates.  I've not had a chance to test these yet on a real FC (except the current one).

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 1]
```diff
@@ -333,42 +333,50 @@ static void _msp_send_altitude(struct msp_bridge *m)
...
+{
+	uint16_t val = manualState.Channel[channel_num];
+	uint16_t min = actuatorSettings.ChannelMin[channel_num];
+	uint16_t max = actuatorSettings.ChannelMax[channel_num];
+	uint16_t idle = actuatorSettings.ChannelNeutral[channel_num];
+
+	// Min and max may be reversed.  If the value is outside of
+	// the given range, return the neutral value.
+	uint16_t lower = min < max ? min : max;
+	uint16_t higher = max > min ? max : min;
```
Comment (mluessi) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: I would use if/else, as the two conditions are the same (`min < max` implies `max > min`).

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 2]
```diff
@@ -333,42 +333,50 @@ static void _msp_send_altitude(struct msp_bridge *m)
...
+	if (val > lower || val < higher) {
+		return idle;
```
Comment (mluessi) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Isn't this always the case? I.e. the value should always be between `lower` and `higher`.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 3]
```diff
@@ -333,42 +333,50 @@ static void _msp_send_altitude(struct msp_bridge *m)
...
+	}
+
+	double d = 1000.0 / ((double)max - (double)min);
```
Comment (mluessi) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Avoid using doubles (they are very expensive on STM32). This should be easily doable using fixed point math (also makes F1 targets happy).

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 4]
```diff
@@ -333,42 +333,50 @@ static void _msp_send_altitude(struct msp_bridge *m)
...
-		MANUALCONTROLCOMMAND_CHANNEL_PITCH,
-		MANUALCONTROLCOMMAND_CHANNEL_YAW,
-		MANUALCONTROLCOMMAND_CHANNEL_THROTTLE,
-		MANUALCONTROLCOMMAND_CHANNEL_FLIGHTMODE,
-		MANUALCONTROLCOMMAND_CHANNEL_ACCESSORY0,
-		MANUALCONTROLCOMMAND_CHANNEL_ACCESSORY1,
-		MANUALCONTROLCOMMAND_CHANNEL_ACCESSORY2,
-	};
+	ActuatorSettingsData actuatorSettings;
+	ActuatorSettingsGet(&actuatorSettings);
```
Comment (mluessi) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: As the settings don't change frequently, you could use a global variable for this and use a callback to update it. This reduces CPU usage but uses more memory (instead using stack, the variable is allocated globally and some memory is used for the callback).

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 2 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: No, I ran into an issue where during failsafe the values are returned out of range.  In the specific case of throttle, this causes the OSD to learn that the range is wider than it'd previously seen and instead of going 0-100%, it goes 0-1%.

I don't know if there's a better way to detect this, but @mlyle suggested just looking at the value to ensure it's in range.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 1 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Thanks, did that.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 3 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: I redid this with fixed-point math.  Seems to do what I expect.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 4 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: OK, I moved this.  Not entirely sure how/when to update it, but it makes sense that the settings won't change much.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 2 above)
Comment (mluessi) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: The code doesn't make sense to be because it assumes `(val > lower || val < higher)` is an exception whereas I would assume it should always be true, i.e., isn't `lower` supposed to be the minimum and `higher` the maximum?

Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: @mluessi Thanks, that was a mistake.  That catch saved me a round trip on my testing.  :)

**Comment by dustin** [comment]
OK.  I spent a bit of time on the bench playing with this.  There was some weirdness about channel reversing I didn't expect (i.e. the min and max values are reversed, but the channel value itself isn't reversed), but this does all the needful.

**Comment by dustin** [comment]
Also added a current consumption change @tracernz pointed out.

**Comment by dustin** [comment]
I pushed up the pre-scaled change, although it's not been tested yet.

Pros:

* A lot less code.
* Less duplication of scaling rules.

Cons:

* Doesn't seem to be a way to send flight mode as an aux channel anymore (without re-scaling).
* Using floating point math now.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 5]
```diff
@@ -218,7 +219,7 @@ static void _msp_send_attitude(struct msp_bridge *m)
 
 	data.att.x = attActual.Roll * 10;
 	data.att.y = attActual.Pitch * -10;
-	data.att.h = 0 * 10;
+	data.att.h = attActual.Yaw;
```
Comment (kubark42) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Something isn't symmetric here. There is a ```*10``` multiplier for Roll and Pitch, but not for yaw. Is this normal?

No matter if it is or isn't, can you please add comments that explain clearly why there are rescaling multipliers?

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 6]
```diff
@@ -283,12 +284,13 @@ static void _msp_send_analog(struct msp_bridge *m)
...
 	}
 
-	// TODO:  Verify these with a board that has working power stuff
 	if (batSettings.VoltagePin != FLIGHTBATTERYSETTINGS_VOLTAGEPIN_NONE)
 		data.status.vbat = (uint8_t)lroundf(batState.Voltage * 10);
 
-	if (batSettings.CurrentPin != FLIGHTBATTERYSETTINGS_CURRENTPIN_NONE)
-		data.status.current = lroundf(batState.Current * 10);
+	if (batSettings.CurrentPin != FLIGHTBATTERYSETTINGS_CURRENTPIN_NONE) {
+		data.status.current = lroundf(batState.Current * 100);
```
Comment (kubark42) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: FWIW, when using variables that are scaled such that they are no longer using "normal" units, my personal preference is to see that in the variable name itself. So this could be data.status.current_100. There's a double advantage to this: 1) it's clear to any downstream users that this variable isn't containing a value that can be used unscaled and 2) if there are any pre-existing downstream users of the variable the compiler will fail until they're all refactored, giving the opportunity to make sure each and every scaling is correct.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 7]
```diff
@@ -333,42 +335,34 @@ static void _msp_send_altitude(struct msp_bridge *m)
...
+		.channels = {
+			_msp_scale_rc(manualState.Roll),
+			_msp_scale_rc(manualState.Pitch),
+			_msp_scale_rc(manualState.Yaw),
+			_msp_scale_rc(manualState.Throttle),
+			_msp_scale_rc(acc0.AccessoryVal),
+			_msp_scale_rc(acc1.AccessoryVal),
+			_msp_scale_rc(acc2.AccessoryVal),
+			1000, // no aux4
+		}};
```
Comment (kubark42) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Style error with ``}}```?

**Comment by kubark42** [comment]
@dustin I don't quite understand the motivation of this PR. Do you think you could flesh out the top post with a little more meat?

P.S. Nice work on having detailed commit messages. :+1:

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 5 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: It's correct, but yeah, that's worthy of a comment.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 8]
```diff
@@ -333,42 +335,34 @@ static void _msp_send_altitude(struct msp_bridge *m)
 	msp_send(m, MSP_ALTITUDE, data.buf, sizeof(data));
 }
 
+static uint16_t _msp_scale_rc(float percent) {
```
Comment (kubark42) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Could you please document this math equation? It will be hard to maintain without understanding its origins.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 6 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: I'm not sure about this one.  It's just a protocol field local to this function and it's changed meaning between a couple versions (and is configurable-ish on the other end).  The protocol docs might be helpful, but it's not well documented there (in particular, the units are documented for voltage, but not current).  It used to be 1/10 of an amp, but at some point some systems changed it to 1/100th of an amp.  Default baseflight is 1/100th, so really *100 is just right for what worked in my out of the box config.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 8 above)
Comment (dustin) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: Added a comment to the top indicating what the goal of this is.

**Comment by kubark42** [comment]
Thanks for the fixes, @dustin. This looks good to me and I'm happy to merge it if you think it's ready. My one question is in regards to ```_msp_scale_rc()``` (which, FWIW, is prefixed with an ```_```, which we don't tend to do in Tau Labs). It seems like the function has evolved a bit since you started, where at first you were looking to scale between the actual max/min PWM, whereas now you are just mapping between 1000 and 2000. Either is fine, I just wanted clarification on this before committing.

**Comment by dustin** [comment]
The output is always expected to be 1000-2000 as that's what MSP wants.  The way I was originally doing it was by reading the config and doing the scaling myself.  It was recommended I use the pre-scaled values and try that.

That latest revision occurred after all my testing, though.  I sent it up because people seemed happier with the concept, but I wouldn't advise sending it through until I've actually seen it work correctly.  It was a little tricky getting all the bits working right on a real board.  MWOSD interprets stick values for some operations and getting the right ones in is a step towards some pretty useful functionality (and immediately provides some anyway).

tl;dr: Please hold off on my final testing of this new change.

**Comment by dustin** [comment]
Just tested this.  Pitch is reversed here (which makes sense, as it's marked as a reversed channel).  That's kind of unfortunate, as I was hoping the reversal would be handled before it hits this scale.  This implies the actual reversal management is happening in other places.

**Comment by mlyle** [comment]
> This implies the actual reversal management is happening in other places.

Nope, we just use a different coordinate system, and as a result our idea of pitch is different from almost everyone else's.  You can confirm this by unreversing the channel and finding that it's now "right" for the OSD.

**Comment by dustin** [comment]
OK.  That sounds good, then.

The next patch will be tested and ready to go:

* Pitch is reversed
* Throttle is scaled from 1100 -> 1900 across ≤0 -> 1

**Comment by dustin** [comment]
OK, this is tested on my sparky.  Everything works as expected.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c (hunk 6 above)
Comment (tracernz) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: The field names don't imply any particular unit and they're not (and I don't imagine ever will be) exported outside the scope of the module. I think within the scope of this module it's reasonable to assume that local variables will be in MultiWii units. To me current_100 is quite meaningless, does it mean this variable contains a value equivalent to 100 A in whatever scaling is being used, or current multiplied by 100, or current divided by 100, or...?

Comment (mlyle) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: yah, I'm kinda meh on this too.  If it was leaving the module some more caution would be warranted.

Comment (kubark42) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: > ... does it mean this variable contains a value equivalent to 100 A in whatever scaling is being used, or current multiplied by 100, or current divided by 100, or...?

That's a fair question and I don't have a way to answer it. What's clear to me, though, is that it's *not* just current and should be investigated more before making assumptions. 

I especially see this kind of scaling error with degrees and radians. Moving to this kind of naming has saved my team tons of time.
